### ISSUE TITLE
Ability to exclude specified StreamField blocks from search

### ISSUE THREAD

**Issue by gasman** [wagtail/wagtail#11127]
### Is your proposal related to a problem?

When indexing a StreamField as part of [search_fields](https://docs.wagtail.org/en/stable/topics/search/indexing.html#indexing-extra-fields), the entire textual content of the StreamField is included. There is no way to specify that a given block type should not be included in the search index, other than by overriding the internal `get_searchable_content` method (which returns a list of strings to be added to the index):

```
from wagtail.blocks import CharBlock

class UnindexedCharBlock(CharBlock):
    def get_searchable_content(self, value):
        return []

# now use UnindexedCharBlock in your StreamField definition in place of CharBlock
```

### Describe the solution you'd like

All blocks that implement a `get_searchable_content` method should be updated to accept a `search_index` argument on their constructor, defaulting to True. If it's set to false, `get_searchable_content` should return an empty list. This will make it possible to write definitions like:

    body = StreamField([
        ('normal_text', blocks.RichTextBlock()),
        ('pull_quote', blocks.RichTextBlock(search_index=False)),
        ('footnotes', blocks.ListBlock(blocks.CharBlock(), search_index=False),
    ], use_json_field=True)

### Implementation pointers

* The `search_index` options should be added to all blocks that implement a `get_searchable_content` method - remember to look in other apps that have a `blocks.py` module, as well as the core `wagtail/blocks` ones. I recommend _not_ adding this on the base `Block` class, since third-party developers may have created their own blocks that don't yet handle the flag in `get_searchable_content`, and it'll be misleading to developers if these accept a `search_index` option that doesn't do anything.
* Adding a new default parameter _might_ have the side effect of generating new migrations for existing (unchanged) Streamfields when the developer next runs `./manage.py makemigrations`. Ideally we don't want to happen - migrations should only be generated when a developer makes a change themselves (such as adding the `search_index=False` flag). You should check this by running `./manage.py makemigrations` on an unmodified copy of the [bakerydemo](https://github.com/wagtail/bakerydemo) project, and verify that no migrations are generated.
* Tests should be added to `wagtail/tests/test_blocks.py`. There are existing tests for `get_searchable_content` in there - hopefully it should be clear how to write new tests for the new option alongside those.
* Documentation: the `search_index` flag should be documented under all of the relevant blocks in `docs/reference/streamfield/blocks.md`. You should also add a new "Search considerations" section somewhere near the bottom of `docs/topics/streamfield.md` explaining the use of `search_index`.

### Additional context

This is a follow-up to the discussion in #7787 - opening this as a new issue for visibility and to single out one specific solution.

**Comment by jgyfutub** [comment]
can I take this up?I am new to wagtail and interested to work upon this issue

**Comment by gasman** [comment]
@jgyfutub Sure, go for it!

**Comment by gasman** [comment]
Fixed in #11135
